network/router.go

```diff
@@ -0,0 +1,306 @@
+	// and will create a new handling routine.
+	err := r.host.Listen(func(c Conn) {
+		dst, err := r.receiveServerIdentity(c)
+		if err != nil {
+			log.Error("receive server identity failed:", err)
+			if err := c.Close(); err != nil {
+				log.Error("Couldn't close secure connection:",
+					err)
+			}
+			return
+		}
+		// pass it along
+		r.launchHandleRoutine(dst, c)
+	})
+	if err != nil {
+		log.Error("Error listening:", err)
+	}
+}
+
+// Stop will stop the listening routine, and stop any routine of handling
+// connections. Calling r.Start(), then r.Stop() then r.Start() again leads to
+// an undefined behaviour. Callers should most of the time re-create a fresh
+// Router.
+func (r *Router) Stop() error {
+	var err error
+	if r.host.Listening() {
+		err = r.host.Stop()
+	}
+	r.connsMut.Lock()
+	// set the isClosed to true
+	r.isClosed = true
+
+	// then close all connections
+	for _, arr := range r.connections {
+		// take all connections to close
+		for _, c := range arr {
+			if err := c.Close(); err != nil {
+				log.Error(err)
+			}
+		}
+	}
+	r.connsMut.Unlock()
+
+	// wait for all handleConn to finish
+	r.wg.Wait()
+
+	r.connsMut.Lock()
+	r.isClosed = false
+	r.connsMut.Unlock()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Send sends to an ServerIdentity without wrapping the msg into a SDAMessage
+func (r *Router) Send(e *ServerIdentity, msg Body) error {
+	if msg == nil {
+		return errors.New("Can't send nil-packet")
+	}
+
+	c := r.connection(e.ID)
+	if c == nil {
+		var err error
+		c, err = r.connect(e)
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Lvlf4("%s sends to %s msg: %+v", r.address, e, msg)
+	var err error
+	err = c.Send(msg)
+	if err != nil {
+		log.Lvl2(r.address, "Couldn't send to", e, ":", err, "trying again")
+		c, err := r.connect(e)
+		if err != nil {
+			return err
+		}
+		err = c.Send(msg)
+		if err != nil {
+			return err
+		}
+	}
+	log.Lvl5("Message sent")
+	return nil
+}
+
+func (r *Router) connect(si *ServerIdentity) (Conn, error) {
+	c, err := r.host.Connect(si.Address)
+	if err != nil {
+		return nil, err
+	}
+	if err := c.Send(r.id); err != nil {
+		return nil, err
+	}
+
+	r.registerConnection(si, c)
+
+	r.launchHandleRoutine(si, c)
+	return c, nil
+
+}
+
+// handleConn is the main routine for a connection to wait for incoming
+// messages.
+func (r *Router) handleConn(remote *ServerIdentity, c Conn) {
+	defer func() {
+		// when leaving, unregister the connection
+		if err := c.Close(); err != nil {
+			log.Error(r.address, "having error closing conn to ", remote.Address, ":", err)
+		}
+		// and release one on the waitgroup
+		r.wg.Done()
+	}()
+	address := c.Remote()
+	log.Lvl3(r.address, "Handling new connection to ", remote.Address)
+	for {
+		packet, err := c.Receive()
+		packet.From = address
+		packet.ServerIdentity = remote
+
+		// whether the router is closed
+		if r.Closed() {
+			// signal we are done with this go routine.
+			return
+		}
+
+		if err != nil {
```

Comment:
Because a `receive` is blocking, so when I close all connections, the `receive` will return and after we check if we are closed or not. If I put the `Closed` before, then some goroutines might leak since the `receive` will never return.